File: aicq-server/src/services/friendRequestService.ts
import { v4 as uuidv4 } from 'uuid';
import { store } from '../db/memoryStore';
import * as friendshipService from './friendshipService';
import type { FriendRequest, FriendPermission } from '../models/types';

/**
 * 发送好友请求
 * - 检查双方账号存在
 * - 不能给自己发
 * - 不能已经是好友
 * - 不能已有 pending 请求（双向检查）
 */
export function sendFriendRequest(
  fromId: string,
  toId: string,
  message?: string,
): FriendRequest {
  if (!fromId || !toId) {
    throw new Error('缺少必填字段: fromId, toId');
  }

  if (fromId === toId) {
    throw new Error('不能给自己发送好友请求');
  }

  // 验证双方账号存在
  const fromNode = store.nodes.get(fromId);
  const toNode = store.nodes.get(toId);
  if (!fromNode) {
    throw new Error('发送者账号不存在');
  }
  if (!toNode) {
    throw new Error('目标账号不存在');
  }

  // 检查是否已经是好友
  if (friendshipService.areFriends(fromId, toId)) {
    throw new Error('你们已经是好友了');
  }

  // 检查是否已有 pending 请求
  for (const req of store.friendRequests.values()) {
    if (req.status !== 'pending') continue;
    // 双向检查：A->B 或 B->A
    if (
      (req.fromId === fromId && req.toId === toId) ||
      (req.fromId === toId && req.toId === fromId)
    ) {
      if (req.fromId === fromId) {
        throw new Error('你已经发送过好友请求，请等待对方回应');
      } else {
        throw new Error('对方已向你发送了好友请求，请先处理该请求');
      }
    }
  }

  const now = Date.now();
  const request: FriendRequest = {
    id: uuidv4(),
    fromId,
    toId,
    status: 'pending',
    message: message?.trim() || undefined,
    createdAt: now,
    updatedAt: now,
  };

  store.friendRequests.set(request.id, request);
  console.log(`[friend-request] ${fromId} 向 ${toId} 发送好友请求 (${request.id})`);
  return request;
}

/**
 * 获取某个账号的所有好友请求（包括发送的和接收的）
 */
export function getFriendRequests(accountId: string): {
  sent: FriendRequest[];
  received: FriendRequest[];
} {
  const sent: FriendRequest[] = [];
  const received: FriendRequest[] = [];

  for (const req of store.friendRequests.values()) {
    if (req.fromId === accountId) {
      sent.push(req);
    } else if (req.toId === accountId) {
      received.push(req);
    }
  }

  // 按时间倒序排列
  sent.sort((a, b) => b.createdAt - a.createdAt);
  received.sort((a, b) => b.createdAt - a.createdAt);

  return { sent, received };
}

/**
 * 接受好友请求
 * - 只能接收者是 toId
 * - 请求必须是 pending 状态
 * - 建立双向好友关系
 * - 根据接受者指定的权限进行授权
 */
export function acceptFriendRequest(
  requestId: string,
  accountId: string,
  permissions?: FriendPermission[],
): FriendRequest {
  const request = store.friendRequests.get(requestId);
  if (!request) {
    throw new Error('好友请求不存在');
  }

  if (request.toId !== accountId) {
    throw new Error('你无权操作此好友请求');
  }

  if (request.status !== 'pending') {
    throw new Error('该请求已被处理');
  }

  // 建立好友关系
  const added = friendshipService.addFriend(request.fromId, request.toId);
  if (!added) {
    request.status = 'rejected';
    request.updatedAt = Date.now();
    store.friendRequests.set(requestId, request);
    throw new Error('添加好友失败，可能是好友数量已达上限');
  }

  // 设置权限：接受者(toId)授予发送者(fromId)的权限
  const defaultPerms: FriendPermission[] = ['chat'];
  const grantedPerms: FriendPermission[] = permissions && permissions.length > 0 ? permissions : defaultPerms;
  // 确保 chat 权限始终存在
  const finalPerms: FriendPermission[] = grantedPerms.includes('chat')
    ? grantedPerms
    : (['chat', ...grantedPerms] as FriendPermission[]);

  request.grantedPermissions = finalPerms;
  request.status = 'accepted';
  request.updatedAt = Date.now();
  store.friendRequests.set(requestId, request);

  // 在 Account 层面也记录权限
  friendshipService.setFriendPermissions(accountId, request.fromId, finalPerms);

  // 初始化对方对己方的默认权限（chat only）
  friendshipService.initDefaultPermissions(request.fromId, accountId);

  console.log(`[friend-request] ${accountId} 接受了 ${request.fromId} 的好友请求 (${requestId})，权限: ${finalPerms.join(', ')}`);
  return request;
}

/**
 * 拒绝好友请求
 * - 只能接收者是 toId
 * - 请求必须是 pending 状态
 */
export function rejectFriendRequest(
  requestId: string,
  accountId: string,
): FriendRequest {
  const request = store.friendRequests.get(requestId);
  if (!request) {
    throw new Error('好友请求不存在');
  }

  if (request.toId !== accountId) {
    throw new Error('你无权操作此好友请求');
  }

  if (request.status !== 'pending') {
    throw new Error('该请求已被处理');
  }

  request.status = 'rejected';
  request.updatedAt = Date.now();
  store.friendRequests.set(requestId, request);

  console.log(`[friend-request] ${accountId} 拒绝了 ${request.fromId} 的好友请求 (${requestId})`);
  return request;
}

File: server/services/friend_request_service.py
"""
AICQ Friend Request Service
=============================
Handles sending, listing, accepting, and rejecting friend requests.
On acceptance, delegates to friendship_service to create the bidirectional
friendship.
"""

from __future__ import annotations

import logging
import uuid
from typing import Any, Dict, List, Optional

from db import DatabaseManager, json_serialize, json_list, iso_now
from config import Config

logger = logging.getLogger("aicq.friend_request_service")


async def send_request(
    db: DatabaseManager,
    from_id: str,
    to_id: str,
    message: Optional[str] = None,
    permissions: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Send a friend request.

    Creates a friend_request with status='pending'.
    Raises ValueError if:
      - from_id and to_id are the same
      - a pending request already exists in either direction
      - the accounts are already friends
    """
    if from_id == to_id:
        raise ValueError("Cannot send a friend request to yourself")

    # Verify both accounts exist
    from_account = await db.fetchone("SELECT id FROM accounts WHERE id = ?", (from_id,))
    if not from_account:
        raise ValueError("Sender account not found")

    to_account = await db.fetchone("SELECT id FROM accounts WHERE id = ?", (to_id,))
    if not to_account:
        raise ValueError("Target account not found")

    # Check for existing pending request in either direction
    existing = await db.fetchone(
        """
        SELECT id FROM friend_requests
        WHERE ((from_id = ? AND to_id = ?) OR (from_id = ? AND to_id = ?))
          AND status = 'pending'
        """,
        (from_id, to_id, to_id, from_id),
    )
    if existing:
        raise ValueError("A pending friend request already exists between these accounts")

    # Check if already friends
    from services.friendship_service import are_friends
    if await are_friends(db, from_id, to_id):
        raise ValueError("These accounts are already friends")

    # Check friend limits
    from_friends_row = await db.fetchone("SELECT friends FROM accounts WHERE id = ?", (from_id,))
    to_friends_row = await db.fetchone("SELECT friends FROM accounts WHERE id = ?", (to_id,))

    from_friend_count = len(json_list(from_friends_row["friends"])) if from_friends_row else 0
    to_friend_count = len(json_list(to_friends_row["friends"])) if to_friends_row else 0

    if from_friend_count >= Config.MAX_FRIENDS:
        raise ValueError("Sender has reached the maximum number of friends")
    if to_friend_count >= Config.MAX_FRIENDS:
        raise ValueError("Target has reached the maximum number of friends")

    if permissions is None:
        permissions = ["chat"]

    request_id = uuid.uuid4().hex
    now = iso_now()

    await db.execute(
        """
        INSERT INTO friend_requests
            (id, from_id, to_id, status, message, granted_permissions, created_at, updated_at)
        VALUES (?, ?, ?, 'pending', ?, ?, ?, ?)
        """,
        (request_id, from_id, to_id, message, json_serialize(permissions), now, now),
    )

    logger.info("Friend request sent: %s -> %s", from_id, to_id)
    return {
        "id": request_id,
        "from_id": from_id,
        "to_id": to_id,
        "status": "pending",
        "message": message,
        "granted_permissions": permissions,
        "created_at": now,
        "updated_at": now,
    }


async def list_requests(
    db: DatabaseManager,
    account_id: str,
) -> Dict[str, List[Dict[str, Any]]]:
    """List sent and received friend requests for an account.

    Returns a dict with 'sent' and 'received' lists.
    """
    sent = await db.fetchall(
        """
        SELECT fr.*, a.display_name as to_display_name, a.agent_name as to_agent_name
        FROM friend_requests fr
        LEFT JOIN accounts a ON fr.to_id = a.id
        WHERE fr.from_id = ?
        ORDER BY fr.created_at DESC
        """,
        (account_id,),
    )

    received = await db.fetchall(
        """
        SELECT fr.*, a.display_name as from_display_name, a.agent_name as from_agent_name
        FROM friend_requests fr
        LEFT JOIN accounts a ON fr.from_id = a.id
        WHERE fr.to_id = ?
        ORDER BY fr.created_at DESC
        """,
        (account_id,),
    )

    # Deserialize JSON permissions
    for req in sent:
        req["granted_permissions"] = json_list(req.get("granted_permissions"))
    for req in received:
        req["granted_permissions"] = json_list(req.get("granted_permissions"))

    return {"sent": sent, "received": received}


async def accept_request(
    db: DatabaseManager,
    request_id: str,
    permissions: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Accept a friend request.

    Updates the request status to 'accepted' and creates a bidirectional
    friendship using the friendship_service.

    Raises ValueError if the request is not found or not in 'pending' status.
    """
    request = await db.fetchone("SELECT * FROM friend_requests WHERE id = ?", (request_id,))
    if not request:
        raise ValueError("Friend request not found")

    if request["status"] != "pending":
        raise ValueError(f"Friend request is already {request['status']}")

    # Use the request's granted_permissions if none provided
    if permissions is None:
        permissions = json_list(request.get("granted_permissions"))
    if not permissions:
        permissions = ["chat"]

    now = iso_now()

    # Update the request status
    await db.execute(
        "UPDATE friend_requests SET status = 'accepted', updated_at = ?, granted_permissions = ? WHERE id = ?",
        (now, json_serialize(permissions), request_id),
    )

    # Create bidirectional friendship
    from services.friendship_service import add_friend_bidirectional
    await add_friend_bidirectional(db, request["from_id"], request["to_id"], permissions)

    logger.info("Friend request accepted: %s (%s -> %s)", request_id, request["from_id"], request["to_id"])

    return {
        "id": request_id,
        "from_id": request["from_id"],
        "to_id": request["to_id"],
        "status": "accepted",
        "granted_permissions": permissions,
        "updated_at": now,
    }


async def reject_request(
    db: DatabaseManager,
    request_id: str,
) -> Dict[str, Any]:
    """Reject a friend request.

    Updates the request status to 'rejected'.
    Raises ValueError if the request is not found or not in 'pending' status.
    """
    request = await db.fetchone("SELECT * FROM friend_requests WHERE id = ?", (request_id,))
    if not request:
        raise ValueError("Friend request not found")

    if request["status"] != "pending":
        raise ValueError(f"Friend request is already {request['status']}")

    now = iso_now()
    await db.execute(
        "UPDATE friend_requests SET status = 'rejected', updated_at = ? WHERE id = ?",
        (now, request_id),
    )

    logger.info("Friend request rejected: %s", request_id)
    return {
        "id": request_id,
        "from_id": request["from_id"],
        "to_id": request["to_id"],
        "status": "rejected",
        "updated_at": now,
    }
